### ISSUE TITLE
Lat/Lon Grid ends within Screen Boundaries, if Map is rotated

### ISSUE THREAD

**Issue by mikebravoyes** [osmdroid/osmdroid#1514]
## Issue Type

[x] Bug
[x] Improvement

## Description and/or steps/code to reproduce the problem

I added the lat/lon grid and the rotation gesture overlays to the map. 
```
// lat/lon grid
LatLonGridlineOverlay2 gridOverlay = new LatLonGridlineOverlay2();
gridOverlay.setMultiplier(4);
gridOverlay.setFontColor(Color.BLACK);
gridOverlay.setBackgroundColor(Color.WHITE);
mapView.getOverlays().add(gridOverlay);

// rotation gesture
RotationGestureOverlay mRotationGestureOverlay;
mRotationGestureOverlay = new RotationGestureOverlay(this, mapView);
mRotationGestureOverlay.setEnabled(true);
mapView.getOverlays().add(mRotationGestureOverlay);
```
Btw: In Android Studio, the RotationGestureOverlay is marked as deprecated, but it works anyway.

When then map is rotated, some of the grid lines end within the screen boundaries - see screenshot below:

![Screenshot_20200308-134951_Sensor Recording](https://user-images.githubusercontent.com/6436590/76163150-1ec32f80-6144-11ea-8dc0-dd54e205d262.jpg)

Independent of the rotation, on small scales (zoom levels), the longitude lines even disappear completely. I can understand, why the latitude is not shown close to polar regions, but the longitude? Please note, that their text labels still exist.

![Screenshot_20200308-135740_Sensor Recording](https://user-images.githubusercontent.com/6436590/76163305-3e0e8c80-6145-11ea-85bf-3cf323711584.jpg)

Not a big problem, but perhaps this could be fixed occasionally.

## Environment

Samsung Galaxy Tab A (2018, 10.5), Android 9, SDK 28 ("Pie")
Also tested on other devices and Android versions - same effect.

### Version of osmdroid the issue relates to:

6.1.6 (in earlier versions as well)

**Comment by spyhunter99** [comment]
definitely a known limitation, solutions are welcome

**Comment by monsieurtanuki** [comment]
@mikebravoyes About [`RotationGestureOverlay`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/gestures/RotationGestureOverlay.java#L25):
````java
/** use {@link #RotationGestureOverlay(MapView)} instead. */
@Deprecated
public RotationGestureOverlay(Context context, MapView mapView) {
	this(mapView);
}
````

@spyhunter99 About [`LatLonGridlineOverlay2.getLatLonGrid`](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/gridlines/LatLonGridlineOverlay2.java#L100):
````java
if (zoom < 2) {
  /*  commented out for performance reasons
...
}
...
//FIXME also draw 2x as wide as the screen, to support rotation?
````
Drawing lines on the canvas wouldn't cost much in terms of performance - and it would be easier to cheat and draw longer lines. Do you remember if there was a specific reason for using `FolderOverlay`s and `Polyline`s?

**Comment by pasniak** [comment]
I also noticed regular `Polyline`s to end prematurely when I rotate the map. It might be relatively new (last .5y?). Wasn't there some poly optimization work?

**Comment by monsieurtanuki** [comment]
@pasniak I don't think it's related: in the current issue the `Polyline` limits explicitly stop at the border of a map with orientation 0 degree, which causes the bug.
If you actually experience Polyline vs. rotation issues, please create a distinct issue with a detailed way to reproduce it.

**Comment by spyhunter99** [comment]
It's definitely in my horrendous logic to draw the grid overlay. There were
a lot of optimizations put in to make it perform well on the v5.2 baseline
(i think that's right). The newer polyline implementations may not need the
optimization and removing it may resolve the issue

**Comment by monsieurtanuki** [comment]
@spyhunter99 I wouldn't call that "horrendous" ;)
I imagined there was a possibility of implementation using the "geodesic" feature, which would definitely explain the use of `Polyline`.
I don't know if the new `Polyline` implementations could make the drawing significantly faster for low zooms: here we're talking about very simple and visible segments, which was not the focus of the latest optimizations.

Besides
* currently in the Lat/Lon grid with `Polyline`s implementation, there are many object creations, which is not recommended in a drawing phase
* there is still the OP issue regarding rotation and ended segment
* there are of course the related infamous "multiworld" side-effects

Working on it...

**Comment by monsieurtanuki** [comment]
Will probably be fixed by the end of the week.

**Comment by monsieurtanuki** [comment]
@mikebravoyes I've just fixed both issues (interrupted segments and no meridian lines on low zoom):
![Screenshot_2020-03-17-16-31-00](https://user-images.githubusercontent.com/11576431/76872740-1918c880-686d-11ea-8564-d15be49e327f.png)

![Screenshot_2020-03-17-16-32-58](https://user-images.githubusercontent.com/11576431/76872746-1c13b900-686d-11ea-9a6f-bd96fdb96c6a.png)

**Comment by monsieurtanuki** [comment]
@mikebravoyes @spyhunter99 Any comment on the position of the labels on the map is welcomed!
For the moment, the labels are centered on the map width or height / 5, on a map with a 0 degree orientation. When rotating, the labels move accordingly - which was more or less the way it worked before.
I couldn't think of something obviously more intuitive.
If it's not a problem I leave it as it is.

**Comment by mikebravoyes** [comment]
@monsieurtanuki 
The position of the labels is OK by me. But the density of the longitudes is a little bit high, so that the picture looks very cluttered with lines and labels.

In the current version 6.1.6 the number of grid lines is mostly between 1 and 3 on my smartphone or between 3 an 5 on my tablet, respectively. This produces a more tidy image.

**Comment by monsieurtanuki** [comment]
@mikebravoyes Thanks for your answer.

I didn't change anything regarding the space between meridians.
Of course the world map on zoom 0 is a bit too much; it's more like a test for world replication.
The big difference between your screenshots and mine are the screens: mine is (only) 480x800 (hdpi). If you take a look at the Netherlands screenshots, we have the same 52N,54N,4E,6E square and about 3x4 lines.
Anyway you can change the grid spacing with a multiplier, display smaller labels or use different colors for the lines and the labels.

I've just merged #1523; feel free to close the issue if there's nothing obvious we missed.
